# "Using Regression for Revenue Prediction of movies"
> "Using TMDb dataset we'll try to predict the revenue of a movie based on the characteristics of the movie, and predict whether a movie's revenue will exceed its budget or not"

- toc: false
- branch: master
- badges: true
- comments: true
- categories: [fastpages, jupyter]
<!-- - image: images/vignette/bird_migration.png -->
- hide: false
- search_exclude: true

Throughout the case study/analysis we'll be using the following libraries:

| Library      | Purpose |
| ----------- | ----------- |
| `sklearn`      | Modelling       |
| `matplotlib`, `bokeh`   | Visualization        |
| `numpy`, `pandas` | Data Manipulation        |

In this case study I am going to do several things first, I want to **predict the revenue** of a movie based on the characteristics of the movie, second I want to **predict whether a movie's revenue will exceed its budget or not**.

# Drudgery: import and take a look at the data

In [41]:
import pandas as pd
import numpy as np

from sklearn.model_selection import cross_val_predict
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import r2_score

import matplotlib.pyplot as plt

df = pd.read_csv("data/processed_data.csv", index_col=0)

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4803 entries, 0 to 4802
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4775 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4391 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4452 non-null   object 
 10  production_countries  4629 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4716 non-null   object 
 15  status               

In [42]:
df.head()

budget                                       genres  \
0  237000000  Action, Adventure, Fantasy, Science Fiction   
1  300000000                   Adventure, Fantasy, Action   
2  245000000                     Action, Adventure, Crime   
3  250000000               Action, Crime, Drama, Thriller   
4  260000000           Action, Adventure, Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash, future, space war, space colony...                en   
1  ocean, drug abuse, exotic island, east india t...                en   
2  spy, based on novel, secret agent, sequel, mi6...                en   
3  dc comics, crime fighter, terrorist, secret id...                en   
4  based on novel, mars, medallion, space travel,...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ...     revenue runtime  \
0  Ingenious Film Partners, Twentieth Century Fox...  ...  2787965087   162.0   
1  Walt Disney Pictures, Jerry Bruckheimer Films,...  ...   961000000   169.0   
2                     Columbia Pictures, Danjaq, B24  ...   880674609   148.0   
3  Legendary Pictures, Warner Bros., DC Entertain...  ...  1084939099   165.0   
4                               Walt Disney Pictures  ...   284139100   132.0   

                                spoken_languages    status  \
0                               English, Español  Released   
1                                        English  Released   
2  Français, English, Español, Italiano, Deutsch  Released   
3                                        English  Released   
4                                        English  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  movie_id  \
0                                    Avatar          7.2      11800     19995   
1  Pirates of the Caribbean: At World's End          6.9       4500       285   
2                                   Spectre          6.3       4466    206647   
3                     The Dark Knight Rises          7.6       9106     49026   
4                               John Carter          6.1       2124     49529   

                                                cast  
0  Sam Worthington, Zoe Saldana, Sigourney Weaver...  
1  Johnny Depp, Orlando Bloom, Keira Knightley, S...  
2  Daniel Craig, Christoph Waltz, Léa Seydoux, Ra...  
3  Christian Bale, Michael Caine, Gary Oldman, An...  
4  Taylor Kitsch, Lynn Collins, Samantha Morton, ...  

[5 rows x 22 columns]

# Data Preprocessing
Our Second step would be to clean and transform the data so that we could apply Regression or Classification algorithm on the data.

## Defining Regression and Classification Outcomes
For regression we'll be using `revenue` as the target for outcomes, and for classification we'll construct an indicator of profitability for each movie. Let's define new column `profitable` such that:
$$
  \begin{equation}
    Profitabe =
    \begin{cases}
      1, & \text{if}\ revenue > budget \\
      0, & \text{otherwise}
    \end{cases}
  \end{equation}
$$

In [43]:
df['profitable'] = df.revenue > df.budget
df['profitable'] = df['profitable'].astype(int)

regression_target = 'revenue'
classification_target = 'profitable'

df['profitable'].value_counts()

1    2585
0    2218
Name: profitable, dtype: int64

2585 out of all movies in the dataset were profitable
## Handling missing and infinite values
Looking at the data we can easily guess that many of the columns are non-numeric and using a technique other than ommiting the columns might be a bit overhead. So I'm going to stick with plane and simple technique of ommiting the column with missing or infinite values.

1. Replace any `np.inf` or `-np.inf` occuring in the dataset with np.nan

In [44]:
df = df.replace([np.inf, -np.inf], np.nan)
print(df.shape)
df.info()

(4803, 23)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 4803 entries, 0 to 4802
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4775 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4391 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4452 non-null   object 
 10  production_countries  4629 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4716 non-null   object 
 15  status    

Notice that `homepage` column accounts for maximun `null` or minimun `non-null` values in the dataset, and we can discard it as a feature for more data.

2. Drop any column with `na` and drop `homepage` column

In [45]:
df.drop('homepage', axis=1, inplace=True)
df = df.dropna(how="any")
df.shape

(3665, 22)

## Transform `genre` column using `OneHotEncoding`

Since `genres` column consists of strings with comma separated genres e.g. `"Action, Adventure, Fantasy"` as a value for a particular movie, I'll convert string to list, then extract all unique genres in the list and finally add a column for each unique genre. Value of a specific genre will be `0` if it is present in `genres` otherwise `0`. 

In [51]:
list_genres = df.genres.apply(lambda x: x.split(","))
genres = []
for row in list_genres:
    row = [genre.strip() for genre in row]
    for genre in row:
        if genre not in genres:
            genres.append(genre)

for genre in genres:
    df[genre] = df['genres'].str.contains(genre).astype(int)

df[genres].head()

Action  Adventure  Fantasy  Science Fiction  Crime  Drama  Thriller  \
0       1          1        1                1      0      0         0   
1       1          1        1                0      0      0         0   
2       1          1        0                0      1      0         0   
3       1          0        0                0      1      1         1   
4       1          1        0                1      0      0         0   

   Animation  Family  Western  Comedy  Romance  Horror  Mystery  History  War  \
0          0       0        0       0        0       0        0        0    0   
1          0       0        0       0        0       0        0        0    0   
2          0       0        0       0        0       0        0        0    0   
3          0       0        0       0        0       0        0        0    0   
4          0       0        0       0        0       0        0        0    0   

   Music  Documentary  Foreign  TV Movie  
0      0            0        0         0  
1      0            0        0         0  
2      0            0        0         0  
3      0            0        0         0  
4      0            0        0         0

## Extract numerical variables
Many of the variables in the dataset are already numerical which will be useful in regression, we'll be extracting these variables and we'll also calculate `skew` of the continuous variables, `plot` these variables.